// web/src/components/DadoComponent.tsx
import React from 'react';
import type { Dado } from '../contexts/GameContext';

interface DadoProps {
  dado: Dado;
  isSelected: boolean;
  onClick: () => void;
  disabled: boolean;
}

const DadoComponent: React.FC<DadoProps> = ({ dado, isSelected, onClick, disabled }) => {
  // Estilo din치mico b치sico
  const estiloDado: React.CSSProperties = {
    cursor: disabled ? 'default' : 'pointer',
    opacity: disabled ? 0.5 : 1,
  };

  // Clases seg칰n tipo y selecci칩n
  const dadoClasses = [
    'dado',
    dado.esCorrupto ? 'corrupto' : 'base',
    isSelected ? 'selected' : '',
  ]
    .filter(Boolean)
    .join(' ');

  const mostrarValor = () => {
    if (dado.valor === null) return '?';
    if (dado.valor === 'CR츼NEO') return '游'; // opcional, m치s claro que texto
    return dado.valor;
  };

  const handleClick = () => {
    if (disabled) return;
    onClick();
  };

  return (
    <div
      className={dadoClasses}
      style={estiloDado}
      onClick={handleClick}
    >
      {mostrarValor()}
    </div>
  );
};

export default DadoComponent;
